### PR TITLE
Add Elrond blockchain

### PR DESCRIPTION
**Pull request by andreibancioiu**
## Description

Extended Wallet-core with blockchain & coin **Elrond**.

## Testing instructions

 * interacting with `walletconsole`
 * running the tests

## Types of changes

* New feature (non-breaking change which adds functionality)

## Checklist

- [ ] Prefix PR title with `[WIP]` if necessary.
- [x] Add tests to cover changes as needed.
- [ ] Update documentation as needed.
- [ ] If there is a related Issue, mention it in the description (e.g. 'Fixes #444 ).

## New blockchain checklist

 - [x] Add the coin definition to `coins.json`.
 - [x] Add coin dispatcher to `src/Coin.cpp`
 - [x] Create tests in `tests/Elrond/TWCoinTypeTests.cpp`
 - [x] Implement `Entry.cpp`
 - [x] Implement `Address.cpp`
 - [ ] ~~Implement `Transaction.cpp` if necessary~~. *Was not necessary*.
 - [x] Implement `Signer.cpp`
 - [x] Write test: mnemonic phrase - > Address derivation test. Put this test in ~~`CoinTests.cpp`~~ `TWHDWalletTests.cpp`.
 - [x] Transaction signing tests, at least one mainnet transaction test.
 - [ ] ~~Add stake, unstake, get rewards tests if the blockchain is PoS like.~~ *Will be added later in time.*
 - [ ] ~~Implement C interface in `src/interface`~~ *Not applicable*.
 - [x] Add tests for `TWAnyAddress` 
 - [x] Add tests for  `TWAnySigner`
 - [x] Add tests in `ElrondTests.swift`
 - [x] Validate generated code in Android project. Write integration tests. *Maiar Android Wallet*.
 - [x] Validate generated code in iOS project. Write integration tests. *Maiar iOS Wallet*.
 - [x] Extend test: `CoinAddressDerivationTests.cpp` 
 - [x] Extend test: `CoinAddressValidationTests.cpp`
 - [x] Extend test: `TWHRPTests.cpp`
 - [x] Extend test `CoinAddressDerivationTests.kt`
 - [x] Extend test `CoinAddressDerivationTests.swift`
 - [ ] Upload coin icon to `trustwallet/assets` if necessary.

### CODE REVIEW

**Comment by hewigovens** [comment]
Hi, have you checked the [integration criteria](https://developer.trustwallet.com/wallet-core/newblockchain#integration-criteria)? pr won't be accepted if not meets

File: docs/coins.md [hunk 1]
```diff
@@ -43,6 +43,7 @@ This list is generated from [./coins.json](../coins.json)
 | 461     | Filecoin         | FIL    | <img src="https://raw.githubusercontent.com/trustwallet/assets/master/blockchains/filecoin/info/logo.png" width="32" />     | <https://filecoin.io/>        |
 | 500     | Theta            | THETA  | <img src="https://raw.githubusercontent.com/trustwallet/assets/master/blockchains/theta/info/logo.png" width="32" />        | <https://www.thetatoken.org>  |
 | 501     | Solana           | SOL    | <img src="https://raw.githubusercontent.com/trustwallet/assets/master/blockchains/solana/info/logo.png" width="32" />       | <https://solana.com>          |
+| 508     | Elrond           | ERD    | <img src="https://raw.githubusercontent.com/trustwallet/assets/master/blockchains/elrond/info/logo.png" width="32" />       | <https://elrond.com/>         |
```
Comment (hewigovens) [inline docs/coins.md:46]: logo.png is 404

File: src/Elrond/Address.h [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+        return Bech32Address::decode(addr, obj_out, hrp);
+    }
+};
+
+} // namespace TW::Elrond
+
+/// Wrapper for C interface.
+struct TWElrondAddress {
+    TW::Elrond::Address impl;
+};
```
Comment (hewigovens) [inline src/Elrond/Address.h:43]: no need
```suggestion
```

File: src/Elrond/Serialization.cpp [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    bool operator() (const string& lhs, const string& rhs) const {
+        return fields_order[lhs] < fields_order[rhs]; 
+    }
+};
+
+template<class Key, class T, class Compare, class Allocator>
+using sorted_map = std::map<Key, T, FieldsSorter, Allocator>;
+using sorted_json = nlohmann::basic_json<sorted_map>;
+
+string TW::Elrond::serializeTransactionToSignableString(const Proto::TransactionMessage& message) {
```
Comment (hewigovens) [inline src/Elrond/Serialization.cpp:37]:
```suggestion
string serializeTransaction(const Proto::TransactionMessage& message) {
```

File: src/Elrond/Serialization.cpp [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+        {"nonce", json(message.nonce())},
+        {"value", json(message.value())},
+        {"receiver", json(message.receiver())},
+        {"sender", json(message.sender())},
+        {"gasPrice", json(message.gas_price())},
+        {"gasLimit", json(message.gas_limit())},
+    };
+
+    if (!message.data().empty()) {
+        payload["data"] = json(TW::Base64::encode(TW::data(message.data())));
```
Comment (hewigovens) [inline src/Elrond/Serialization.cpp:48]:
```suggestion
        payload["data"] = json(Base64::encode(TW::data(message.data())));
```

File: src/Elrond/Serialization.cpp [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+    }
+
+    return payload.dump();
+}
+
+string TW::Elrond::serializeSignedTransaction(const Proto::TransactionMessage& message, string signature) {
```
Comment (hewigovens) [inline src/Elrond/Serialization.cpp:54]:
```suggestion
string serializeSignedTransaction(const Proto::TransactionMessage& message, string signature) {
```

File: src/Elrond/Signer.h [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+    /// Signs a json Proto::SigningInput with private key
+    static std::string signJSON(const std::string& json, const Data& key);
+};
+
+} // namespace TW::Elrond
+
+/// Wrapper for C interface.
+struct TWElrondSigner {
+    TW::Elrond::Signer impl;
+};
```
Comment (hewigovens) [inline src/Elrond/Signer.h:33]:
```suggestion
```

File: src/Elrond/Serialization.cpp [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+    sorted_json payload {
+        {"nonce", json(message.nonce())},
+        {"value", json(message.value())},
+        {"receiver", json(message.receiver())},
+        {"sender", json(message.sender())},
+        {"gasPrice", json(message.gas_price())},
+        {"gasLimit", json(message.gas_limit())},
+        {"data", json(message.data())},
```
Comment (hewigovens) [inline src/Elrond/Serialization.cpp:62]: looks duplicate code in `serializeTransactionToSignableString`, also, `data` is not checked?

File: src/Elrond/Signer.cpp [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+#include "Serialization.h"
+#include "../PublicKey.h"
+#include "HexCoding.h"
+
+#include <google/protobuf/util/json_util.h>
+
+using namespace TW;
+using namespace TW::Elrond;
+
+
```
Comment (hewigovens) [inline src/Elrond/Signer.cpp:18]:
```suggestion
```

File: tests/Elrond/TestAccounts.h [hunk 9]
```diff
@@ -0,0 +1,11 @@
+#pragma once
```
Comment (hewigovens) [inline tests/Elrond/TestAccounts.h:1]: missing copyright header

File: tests/Elrond/TWCoinTypeTests.cpp [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+//
+
+#include "../interface/TWTestUtilities.h"
+#include <TrustWalletCore/TWCoinTypeConfiguration.h>
+#include <gtest/gtest.h>
+
+
+TEST(TWElrondCoinType, TWCoinType) {
+    auto symbol = WRAPS(TWCoinTypeConfigurationGetSymbol(TWCoinTypeElrond));
+    auto txId = TWStringCreateWithUTF8Bytes("t123");
```
Comment (hewigovens) [inline tests/Elrond/TWCoinTypeTests.cpp:18]: please use real world tx id if possible

File: tests/Elrond/TWCoinTypeTests.cpp [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+    auto txUrl = WRAPS(TWCoinTypeConfigurationGetTransactionURL(TWCoinTypeElrond, txId));
+    auto accId = TWStringCreateWithUTF8Bytes("a12");
```
Comment (hewigovens) [inline tests/Elrond/TWCoinTypeTests.cpp:20]: same here

File: src/proto/Elrond.proto [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+package TW.Elrond.Proto;
+option java_package = "wallet.core.jni.proto";
+
+// A transaction, typical balance transfer
+message TransactionMessage {
+    uint64      nonce = 1;
+    string      value = 2;
+    string      receiver = 3;
+    string      sender = 4;
+    uint64      gas_price = 5;
```
Comment (hewigovens) [inline src/proto/Elrond.proto:18]: Elrond decimals is 18, overflow if you use type `uint64`?

File: src/proto/Elrond.proto [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+    bytes private_key = 1;
+
+    oneof message_oneof {
+        TransactionMessage transaction = 2;
+    }
+}
+
+// Transaction signing output.
+message SigningOutput {
+    string signed_transaction = 1;
```
Comment (hewigovens) [inline src/proto/Elrond.proto:34]:
```suggestion
    string encoded = 1;
```

File: swift/Tests/Blockchains/ElrondTests.swift [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+                $0.data = "foo"
+            }
+
+            $0.privateKey = privateKey.data
+        }
+
+        let output: ElrondSigningOutput = AnySigner.sign(input: input, coin: .elrond)
+        let expectedSignature = "b88ad2fe98a7316ea432a0a76c18cd87200fe75f27a8f053ea6532b40317dbec5136c5463aef132ae951b7e60d45d921caaa5903e70821dcda98f237d4ec4308"
+
+        XCTAssertEqual(output.signature, expectedSignature)
```
Comment (hewigovens) [inline swift/Tests/Blockchains/ElrondTests.swift:47]: can you also test `signed_transaction`?

File: tests/Elrond/SignerTests.cpp [hunk 15]
```diff
@@ -0,0 +1,89 @@
...
+    input.mutable_transaction()->set_receiver(BOB_BECH32);
+    input.mutable_transaction()->set_gas_price(200000000000000);
+    input.mutable_transaction()->set_gas_limit(500000000);
+    input.mutable_transaction()->set_data("foo");
+
+    auto output = Signer::sign(input);
+    auto signature = output.signature();
+    auto signedTransaction = output.signed_transaction();
+    auto expectedSignature = "b88ad2fe98a7316ea432a0a76c18cd87200fe75f27a8f053ea6532b40317dbec5136c5463aef132ae951b7e60d45d921caaa5903e70821dcda98f237d4ec4308";
+    auto expectedSignedTransaction = (boost::format(R"({"nonce":0,"value":"0","receiver":"%1%","sender":"%2%","gasPrice":200000000000000,"gasLimit":500000000,"data":"foo","signature":"%3%"})") % BOB_BECH32 % ALICE_BECH32 % expectedSignature).str();
```
Comment (hewigovens) [inline tests/Elrond/SignerTests.cpp:38]: is this a valid tx in block explorer?

File: src/proto/Elrond.proto (hunk 12 above)
Comment (andreibancioiu) [reply src/proto/Elrond.proto:18]: Very good question. However we expect `gas_price` to be much, much lower than `max(uint64)` for any transaction, with respect to our network's economics. The `gas_price`, chosen by the issuer of the transaction, is per gas unit - while `gas_limit` (the number of gas units) is constrained to have a quite high (~50000) lower bound, and thus a very large `gas_price` would be prohibitive (balance exhaustion). 

Once the transaction reaches the network, we use `big numbers` to perform calculations on this matter.

File: docs/coins.md (hunk 1 above)
Comment (andreibancioiu) [reply docs/coins.md:46]: Indeed, we did not open a pull request against the repository `trustwallet/assets` yet.

File: tests/Elrond/TestAccounts.h (hunk 9 above)
Comment (andreibancioiu) [reply tests/Elrond/TestAccounts.h:1]: Added header.

File: tests/Elrond/TWCoinTypeTests.cpp (hunk 10 above)
Comment (andreibancioiu) [reply tests/Elrond/TWCoinTypeTests.cpp:18]: Fixed.

File: tests/Elrond/TWCoinTypeTests.cpp (hunk 11 above)
Comment (andreibancioiu) [reply tests/Elrond/TWCoinTypeTests.cpp:20]: Fixed.

File: src/Elrond/Serialization.cpp (hunk 7 above)
Comment (andreibancioiu) [reply src/Elrond/Serialization.cpp:62]: There is duplication, indeed - a bit closer to _accidental_ duplication though, and we would like to keep `serializeTransaction` (called `serializeTransactionToSignableString` before review) separately.

Indeed, the `data` field is applied a special treatment in the first serialization method, whose output is meant to be signed - and we have a more constrained procedure here, such as: omit the data field if empty, apply _base64_ on it. On the other hand, these constraints do not apply to obtaining the actual message that should be dispatched to the blockchain.

Is it acceptable on your side if we keep these as they are?

File: swift/Tests/Blockchains/ElrondTests.swift (hunk 14 above)
Comment (andreibancioiu) [reply swift/Tests/Blockchains/ElrondTests.swift:47]: Thank you for noticing. Added the test.

File: tests/Elrond/SignerTests.cpp (hunk 15 above)
Comment (andreibancioiu) [reply tests/Elrond/SignerTests.cpp]: Yes, it is - the encoded transaction to be sent to the blockchain would look like this:

```
{
    "nonce": 0,
    "value": "0",
    "sender": "erd1l453hd0gt5gzdp7czpuall8ggt2dcv5zwmfdf3sd3lguxseux2fsmsgldz",
    "receiver": "erd1cux02zersde0l7hhklzhywcxk4u9n4py5tdxyx7vrvhnza2r4gmq4vw35r",
    "gasPrice": 200000000000000,
    "gasLimit": 500000000,
    "data": "foo",
    "signature": "b88ad2fe98a7316ea432a0a76c18cd87200fe75f27a8f053ea6532b40317dbec5136c5463aef132ae951b7e60d45d921caaa5903e70821dcda98f237d4ec4308"
}
```

The computed hash (returned upon invoking the API) is `cbecce3712064fe8b14bd668f02cf348d4ee5e4c73836d0a7e25482e5d0056d9`, and the link in block explorer would be:

https://explorer.elrond.com/transactions/cbecce3712064fe8b14bd668f02cf348d4ee5e4c73836d0a7e25482e5d0056d9

The link above is currently available (and will be until a hard reset of our testnet).

Does this answer the question?

**Comment by andreibancioiu** [comment]
@hewigovens thanks a lot for your review! We have applied the fixes and added some responses.

**Pull request merged**